yapapi/executor/__init__.py

```diff
@@ -566,6 +571,33 @@ async def execute_tasks(
             async for t in executor.submit(worker, data):
                 yield t
 
+    async def run_service(
+        self,
+        service_class: Type["Service"],
+        num_instances: int = 1,
+        payload: Optional[Payload] = None,
+        expiration: Optional[datetime] = None,
+    ) -> "Cluster":
+        from .services import Cluster  # avoid circular dependency
+
+        payload = payload or await service_class.get_payload()
+
+        if not payload:
+            raise ValueError(
+                f"No payload returned from {service_class.__name__}.get_payload() nor given in the `payload` argument."
+            )
+
+        cluster = Cluster(
+            engine=self,
+            service_class=service_class,
+            payload=payload,
+            num_instances=num_instances,
+            expiration=expiration,
+        )
+        await self._stack.enter_async_context(cluster)
```

Comment:
Managing `cluster`'s context this way is a bit problematic. It means that `cluster.__aexit__()` is called when the `Golem` instance referred to by `self` in this code calls `__aexit__()`. The current implementation of `Golem.__aexit__()` is as follows (most details omitted):
```python
async def Golem.__aexit__():
   wait for all jobs to finish
   call self._stack.exit()  # this will call cluster.__aexit__()
```
And `Cluster.__aexit__()` looks like follows:
```python
async def Cluster.__aexit__():
    finalize the job
```
So `Golem.__aexit__()` is stuck waiting for the job to be marker finished, which doesn't happen, since the job is finialized later, when `cluster.__aexit__()` is called.

Reply:
fixed in https://github.com/golemfactory/yapapi/pull/383/commits/11a3655ab70afe8e3106f8147b3c9d85679812f4

Reply:
<3